hazelcast/src/test/java/com/hazelcast/spi/impl/operationservice/impl/InvocationNestedLocalTest.java

```diff
@@ -8,115 +8,120 @@
         InternalCompletableFuture f = operationService.invokeOnPartition(null, outerOperation, outerOperation.getPartitionId());
 
-        assertEquals(response, f.getSafely());
+        assertEquals(RESPONSE, f.getSafely());
     }
 
     @Test
-    public void whenPartition_callsCorrectPartition() {
-        HazelcastInstance hz = createHazelcastInstance();
-        OperationService operationService = getOperationService(hz);
+    public void invokeOnPartition_outerLocal_innerSameInstance_samePartition() {
+        HazelcastInstance local = createHazelcastInstance();
+        OperationService operationService = getOperationService(local);
 
-        int partitionId = 0;
-        InnerOperation innerOperation = new InnerOperation(response, partitionId);
+        int partitionId = getPartitionId(local);
+        InnerOperation innerOperation = new InnerOperation(RESPONSE, partitionId);
         OuterOperation outerOperation = new OuterOperation(innerOperation, partitionId);
-
         InternalCompletableFuture f = operationService.invokeOnPartition(null, outerOperation, outerOperation.getPartitionId());
 
-        assertEquals(response, f.getSafely());
+        assertEquals(RESPONSE, f.getSafely());
     }
 
     @Test
-    public void whenPartition_callsIncorrectPartition() {
-        HazelcastInstance hz = createHazelcastInstance();
-        OperationService operationService = getOperationService(hz);
-
-        int outerPartitionId = 0;
-        int innerPartitionId = 1;
-        for (; innerPartitionId < hz.getPartitionService().getPartitions().size(); innerPartitionId++) {
-            if (!mappedToSameThread(operationService, outerPartitionId, innerPartitionId)) {
-                break;
-            }
-        }
-
-        InnerOperation innerOperation = new InnerOperation(response, innerPartitionId);
-        OuterOperation outerOperation = new OuterOperation(innerOperation, outerPartitionId);
+    public void invokeOnPartition_outerLocal_innerSameInstance_callsDifferentPartition() {
+        HazelcastInstance local = createHazelcastInstance();
+        OperationService operationService = getOperationService(local);
 
+        int outerPartitionId = getPartitionId(local);
+        int innerPartitionId = randomPartitionIdNotMappedToSameThreadAsGivenPartitionIdOnInstance(local, outerPartitionId);
+        InnerOperation innerOperation = new InnerOperation(RESPONSE, innerPartitionId);
+        OuterOperation outerOperation = new OuterOperation(innerOperation, outerPartitionId);
         InternalCompletableFuture f = operationService.invokeOnPartition(null, outerOperation, outerOperation.getPartitionId());
 
-        try {
-            f.getSafely();
-            fail();
-        } catch (IllegalThreadStateException e) {
-        }
+        expected.expect(IllegalThreadStateException.class);
+        expected.expectMessage("cannot make remote call");
```

Comment:
I know there are various opinions about it - I wouldn't like to have another dogmatic discussion about it and would prefer to stay pragmatic.
We don't have any error codes, thus it's difficult to assert precisely what you want to assert. 
Checking exception type only is in many cases very imprecise, you can also get:
`cannot make invocation on multiple partitions`
`cannot be run in current thread`

I prefer to be sure what I'm asserting and where the exception is thrown - otherwise the test may not be testing what you are willing to test. I've seen many of them already...